homestead.root/usr/share/clearwater/bin/poll_cx.py

```diff
@@ -0,0 +1,287 @@
+# copying, modification, propagation, or distribution would otherwise
+# violate the terms of the GPL. You must comply with the GPL in all
+# respects for all of the code used other than OpenSSL.
+# "OpenSSL" means OpenSSL toolkit software distributed by the OpenSSL
+# Project and licensed under the OpenSSL Licenses, or a work based on such
+# software and licensed under the OpenSSL Licenses.
+# "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
+# under which the OpenSSL Project distributes the OpenSSL toolkit software,
+# as those licenses appear in the file LICENSE-OPENSSL.
+
+
+# This poll script determines whether, during the "Previous 5 minutes" stat
+# period, more than a configured threshold of Diameter request got rejected
+# with "server" errors (i.e. errors which would have prompted Homestead to
+# respond to the caller with a 5XX HTTP code) and NO requests succeeded over the
+# same period.  monit responds to a positive response from this script by
+# restarting homestead.
+#
+# The rationale for this function is that, if many diameter requests are failing
+# with server errors and no requests are succeeding, this _might_ be because
+# the Diameter stack is broken (rather than a genuine server or configuration
+# issue) perhaps because freeDiameter failed to handle an unsual network
+# failure and "hung" the Diameter connections.  In cases such as this, it is
+# likely that service can be restored by restarting homestead, and there is no
+# harm in trying this, as no requests are succeeding anyway.
+#
+# Note that the script does the stat lookup and calculation twice (and only
+# responds in the positive if both checks suggest that the stack might be
+# broken).  The reasoning here is that its not possible to check the "Previous
+# 5 minute" value for all statistics atomically and therefore the values of
+# some stats might be for a different 5 minute period.  If the script only
+# checked once, it would be vulnerable to this "tick rollover" problem and
+# might incorrectly identify the stack as needing a reboot, invalidly
+# impacting service.
+
+import sys
+import re
+import netsnmp
+import operator
+import logging
+
+# OID template for that segment of a Cx stat table covering previous 5 minute
+# counts
+table_segment_head = '.1.2.826.0.1.1578918.9.5.%d.1.4.3'
+
+#logging.basicConfig(level=logging.DEBUG)
+_log = logging.getLogger(__name__)
+
+# Get tolerance to server errors. We will only treat the number of server
+# errors at the Diameter interface as serious enough to warrant a restart
+# (provided there have been no successes over the same period) if the total
+# number of errors exceeds this tolerance.
+server_error_tolerance = int(sys.argv[1])
+_log.info("Server error tolerance = {}".format(server_error_tolerance))
+
+# Get a connection to the local SNMP agent
+ss = netsnmp.Session(DestHost='localhost', Version=2, Community='clearwater')
```

Comment:
This should be `clearwater-internal` (see [here](https://github.com/Metaswitch/clearwater-infrastructure/blob/6ffb0f6a021b78845583ce2d1e6b882becdcec6d/clearwater-snmpd/etc/snmp/snmpd.conf.clearwater-snmpd#L247)) - for security, users can change the community string that's accessible from other boxes from `clearwater` to something else.